src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheService.cs

```diff
@@ -0,0 +1,140 @@
+            }
+        }
+
+        public void OnAnalyzeDocument(Document document)
+        {
+            var isAnalyzedForFirstTime = OnLiveAnalysisStarted(document);
+            if (isAnalyzedForFirstTime && document.IsOpen())
+            {
+                // Only need to do this once, i.e. only the first time, in case there no diagnostics in this document,
+                // then we want to make sure clear cached diagnostics for this document, after that it will be handled
+                // by monitoring diagnostics update.
+                _cacheUpdater.QueueUpdate(document.Id);
+            }
+        }
+
+        private static async Task<ImmutableArray<DiagnosticData>> GetCachedDiagnosticsAsync(Document document, CancellationToken cancellationToken)
+        {
+            var client = await RemoteHostClient.TryGetClientAsync(document.Project.Solution.Workspace, cancellationToken).ConfigureAwait(false);
+            if (client == null)
+            {
+                return default;
+            }
+
+            var checksum = await GetChecksumAsync(document, cancellationToken).ConfigureAwait(false);
+
+            var result = await client.TryInvokeAsync<IRemoteDiagnosticCacheService, ImmutableArray<DiagnosticData>>(
+                (service, ct) => service.GetCachedDiagnosticsAsync((DocumentKey.ToDocumentKey(document)).Dehydrate(), checksum, ct),
+                cancellationToken).ConfigureAwait(false);
+
+            if (result.HasValue)
+            {
+                var diagnostics = result.Value.SelectAsArray(FixDescription);
+                return diagnostics;
+            }
+
+            return default;
+
+            // Change text to make it clear this is loaded from cache.
+            static DiagnosticData FixDescription(DiagnosticData d) =>
+                new(id: d.Id, category: d.Category,
+                    message: string.Format(CachedMessageFormat, d.Message),
+                    enuMessageForBingSearch: d.ENUMessageForBingSearch,
+                    severity: d.Severity, defaultSeverity: d.DefaultSeverity, isEnabledByDefault: d.IsEnabledByDefault, warningLevel: d.WarningLevel,
+                    customTags: d.CustomTags, properties: d.Properties, projectId: d.ProjectId, location: d.DataLocation, additionalLocations: d.AdditionalLocations,
+                    language: d.Language, title: d.Title,
+                    description: d.Description,
+                    helpLink: d.HelpLink, isSuppressed: d.IsSuppressed);
+        }
+
+        private static async ValueTask<Checksum> GetChecksumAsync(Document document, CancellationToken cancellationToken)
+        {
+            // We only checksum off of the contents of the file.  During load, we can't really compute any other
+            // information since we don't necessarily know about other files, metadata, or dependencies.  So during
+            // load, we allow for the previous diagnostics to be used as long as the file contents match.
```

Comment:
This part feels a bit scary - if there were some errors in the prior session due to things like missing references or packages, and the user closed VS, restored and built everything successfully on command line and then re-opened VS, they would still see stale errors from previous session until we replace cached diagnostics with true live diagnostics.

Reply:
Yeah, this is a very fair concern, any project level change might be problematic in current design. Additionally, I think we could end up caching "wrong" diagnostics in IDE caused by incorrect project restore, which user fixed in commandline w/o changes to projects. I don't have a good solution to this at the moment.

Reply:
@vatsalyaagrawal ^ and https://github.com/dotnet/roslyn/pull/50168#issuecomment-784442326